File: January/Day1/word_pattern.cpp
//https://leetcode.com/problems/word-pattern/description/

class Solution {
public:
   void split(string &s,vector<string>&ans) 
{
    istringstream ss(s);
 
    string word;
    while (ss >> word)
    {
    ans.push_back(word);
    }
}
    bool wordPattern(string pattern, string s) {
    unordered_map<char,string>mp;
    vector<string>v;
    unordered_set<string>taken;
    split(s, v);
    if(pattern.size()!=v.size())
    {
        return false;
    }
    for(int i=0;i<pattern.size();i++)
    {
      if(mp.find(pattern[i])==mp.end() && taken.find(v[i])== taken.end()) 
      {
      mp[pattern[i]]=v[i];
      taken.insert(v[i]);
      } 
      else
      {
          if(mp[pattern[i]]!=v[i])
          {
              return false;
          }
      }
    }
return true;
    }
};

File: January/Day1/word_pattern.java

class Solution {
    public boolean wordPattern(String pattern, String s) {
        HashMap<Character,String> map = new HashMap<>();

        String[] str = s.split(" ");

        if(str.length!=pattern.length())
            return false;

        for(int i=0;i<pattern.length();i++)
        {
              
            char ch = pattern.charAt(i);

            if(map.containsKey(ch)){
                if(!map.get(ch).equals(str[i])){
                    return false;
                }
                    
            }
            else
            { 
              if(map.containsValue(str[i]))
                return false;
              map.put(ch,str[i]);
            }
          

        }
       
        return true;

    }
}
